ksql-engine/src/main/java/io/confluent/ksql/function/udf/array/Entries.java

```diff
@@ -0,0 +1,109 @@
+/*
+ * Copyright 2019 Confluent Inc.
+ *
+ * Licensed under the Confluent Community License (the "License"; you may not use
+ * this file except in compliance with the License. You may obtain a copy of the
+ * License at
+ *
+ * http://www.confluent.io/confluent-community-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.ksql.function.udf.array;
+
+import io.confluent.ksql.function.udf.Udf;
+import io.confluent.ksql.function.udf.UdfDescription;
+import io.confluent.ksql.function.udf.UdfParameter;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import org.apache.kafka.connect.data.Schema;
+import org.apache.kafka.connect.data.SchemaBuilder;
+import org.apache.kafka.connect.data.Struct;
+
+/**
+ * This UDF constructs an array of structs from the entries in a map. Each struct has a field with
+ * name "K" containing the key (this is always a String) and a field with name "V" holding the
+ * value;
+ */
+@UdfDescription(name = "ENTRIES", description = "Construct an array from the entries in a map")
+public class Entries {
+
+  private static final Schema INT_STRUCT_SCHEMA = buildStructSchema(Schema.OPTIONAL_INT32_SCHEMA);
+  private static final Schema BIGINT_STRUCT_SCHEMA = buildStructSchema(
+      Schema.OPTIONAL_INT64_SCHEMA);
+  private static final Schema DOUBLE_STRUCT_SCHEMA = buildStructSchema(
+      Schema.OPTIONAL_FLOAT64_SCHEMA);
+  private static final Schema BOOLEAN_STRUCT_SCHEMA = buildStructSchema(
+      Schema.OPTIONAL_BOOLEAN_SCHEMA);
+  private static final Schema STRING_STRUCT_SCHEMA = buildStructSchema(
+      Schema.OPTIONAL_STRING_SCHEMA);
+  private static final String KEY_FIELD_NAME = "K";
+  private static final String VALUE_FIELD_NAME = "V";
+
+  private static Schema buildStructSchema(final Schema valueSchema) {
+    return SchemaBuilder.struct().field(KEY_FIELD_NAME, Schema.OPTIONAL_STRING_SCHEMA)
+        .field(VALUE_FIELD_NAME, valueSchema).optional().build();
+  }
+
+  @Udf(schema = "ARRAY<STRUCT<K STRING, V INT>>")
```

Comment:
I feel like this might be a good use case for schema provider - for now it's manageable because we only support string keys but if we ever support anything else this will be combinatorial! 

this would also prevent us from needing all the constants up top.

Reply:
I don't know how you would do this with schema provider. We need to know the schema at runtime here in order to construct the structs properly. AIUI there is no way to get the schema from inside the actual udf function.